internal/app/machined/pkg/system/services/etcd.go

```diff
@@ -0,0 +1,331 @@
+	"github.com/talos-systems/talos/pkg/net"
+	"github.com/talos-systems/talos/pkg/userdata"
+)
+
+var etcdImage = fmt.Sprintf("%s:%s", constants.EtcdImage, constants.DefaultEtcdVersion)
+
+// Etcd implements the Service interface. It serves as the concrete type with
+// the required methods.
+type Etcd struct{}
+
+// ID implements the Service interface.
+func (e *Etcd) ID(data *userdata.UserData) string {
+	return "etcd"
+}
+
+// PreFunc implements the Service interface.
+func (e *Etcd) PreFunc(ctx context.Context, data *userdata.UserData) (err error) {
+	if _, err = os.Stat("/etc/kubernetes/pki/etcd"); os.IsNotExist(err) {
+		if err = generatePKI(data); err != nil {
+			return errors.Wrap(err, "failed to generate etcd PKI")
+		}
+	}
+
+	client, err := containerdapi.New(constants.ContainerdAddress)
+	if err != nil {
+		return err
+	}
+	// nolint: errcheck
+	defer client.Close()
+
+	// Pull the image and unpack it.
+	containerdctx := namespaces.WithNamespace(ctx, constants.SystemContainerdNamespace)
+	if _, err = client.Pull(containerdctx, etcdImage, containerdapi.WithPullUnpack); err != nil {
+		return fmt.Errorf("failed to pull image %q: %v", etcdImage, err)
+	}
+
+	return nil
+}
+
+// PostFunc implements the Service interface.
+func (e *Etcd) PostFunc(data *userdata.UserData) (err error) {
+	return nil
+}
+
+// Condition implements the Service interface.
+func (e *Etcd) Condition(data *userdata.UserData) conditions.Condition {
+	if data.Service.Kubeadm.IsControlPlane() {
+		return conditions.WaitForFileToExist("/etc/kubernetes/kubelet.conf")
+	}
+
+	return nil
+}
+
+// DependsOn implements the Service interface.
+func (e *Etcd) DependsOn(data *userdata.UserData) []string {
+	if data.Service.Kubeadm.IsControlPlane() {
+		return []string{"containerd", "kubelet"}
+	}
+
+	return []string{"containerd"}
+}
+
+// Runner implements the Service interface.
+func (e *Etcd) Runner(data *userdata.UserData) (runner.Runner, error) {
+	ips, err := net.IPAddrs()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to discover IP addresses")
+	}
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+
+	initialClusterState := "new"
+	initialCluster := ""
+	if data.Services.Kubeadm.IsBootstrap() {
+		initialCluster = "--initial-cluster=" + hostname + "=https://" + ips[0].String() + ":2380"
+	} else if data.Services.Kubeadm.IsControlPlane() {
+		h, err := kubernetes.NewHelper()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create client")
+		}
+
+		endpoints, err := h.MasterIPs()
+		if err != nil {
+			return nil, err
+		}
+
+		resp, err := addMember(endpoints)
+
+		for _, member := range resp.Members {
+			initialCluster = fmt.Sprintf("%s,%s", initialCluster, member.ClientURLs[0])
+		}
+	}
+
+	// Set the process arguments.
+	args := runner.Args{
+		ID: e.ID(data),
+		ProcessArgs: []string{
+			"/usr/local/bin/etcd",
+			"--name=" + hostname,
+			"--listen-client-urls=https://0.0.0.0:2380",
```

Comment:
We should probably make a package which takes an IP address (ips[0] in this case), checks whether it is IPv4 or IPv6, and offers AllAddress() (0.0.0.0 or ::) and LoopBackAddress() (127.0.0.1 or ::1) for each.

Reply:
Good point 👍

Reply:
@Ulexus gonna follow up with a IPv6/IPV4 dual stack helper library.